### PR TITLE
[agg_v2] Fix mistakenly overriding on concurrent set_base_value

### PR DESCRIPTION
**Pull request by igor-aptos**

### CODE REVIEW

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 1]
```diff
@@ -77,6 +77,30 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> Default
...
+        &mut self,
+        shifted_idx: ShiftedTxnIndex,
+        incarnation: Incarnation,
+        // TODO add layout to values
+        values: impl IntoIterator<Item = (T, V)>,
+    ) {
+        match self.idx_to_update.get(&shifted_idx) {
+            Some(previous) => {
+                // base value may have already been provided due to a concurrency race.
+                // If maybe_layout is None, they are required to be identical.
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:90]: The comment is wrong / old / stale?

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 2]
```diff
@@ -77,6 +77,30 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> Default
...
+                // If maybe_layout is Some, there might have been an exchange
+                // Assert the length of bytes for efficiency (instead of full equality)
+                for (tag, v) in values.into_iter() {
+                    let prev_v = previous
+                        .get(&tag)
+                        .expect("Reading twice from storage must be consistent");
+                    assert!(v.bytes_len() == prev_v.bytes_len());
+                }
+            },
+            None => self.write(shifted_idx, incarnation, values),
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:100]: What is the problem with overwriting concurrently the base values? They should be identical?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 3]
```diff
@@ -259,21 +264,41 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
...
-                // base value may have already been provided due to a concurrency race,
-                // but it has to be the same as being set.
-                // Assert the length of bytes for efficiency (instead of full equality)
-                *i == 0 && v.bytes_len() == bytes_len
-            } else {
-                true
-            }
-        }));
+        use btree_map::Entry::*;
+        match v.versioned_map.entry(ShiftedTxnIndex::zero()) {
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_data.rs:269]: Why match over old code? Seems like layout is the only thing that got added?

Comment (igor-aptos) [reply aptos-move/mvhashmap/src/versioned_data.rs:269]: we cannot do insert - that overwrites existing value. we need to do insert only if Vacant.

https://aptos-org.slack.com/archives/C057VPCD7QX/p1697520278313169

File: aptos-move/mvhashmap/src/versioned_group_data.rs (hunk 1 above)
Comment (igor-aptos) [reply aptos-move/mvhashmap/src/versioned_group_data.rs:90]: it is future looking! 
once we add maybe layout (TODO above), that is what check should be doing :D

File: aptos-move/mvhashmap/src/versioned_group_data.rs (hunk 2 above)
Comment (igor-aptos) [reply aptos-move/mvhashmap/src/versioned_group_data.rs:100]: not if there were aggregators, and they got exchanged with ephemeral IDs

**Review by georgemitenkov — APPROVED** [review]

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 4]
```diff
@@ -77,6 +77,30 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> Default
 }
 
 impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> VersionedGroupValue<T, V> {
+    fn set_base_values(
+        &mut self,
+        shifted_idx: ShiftedTxnIndex,
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:82]: What does shifted transaction index mean here? Why is transaction index shifted?

Comment (igor-aptos) [reply aptos-move/mvhashmap/src/versioned_group_data.rs:82]: we need to distinguish storage value from all other transactions.

So shifted means 0 is storage, and i>0 means transaction i-1

I forgot the reason why we didn't do Option<TxnIndex> instead though (I mentioned but @gelash had a reason)

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 5]
```diff
@@ -259,21 +264,41 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
...
+            Vacant(v) => {
+                v.insert(CachePadded::new(Entry::new_write_from(
+                    0,
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/versioned_data.rs:272]: Is it okay to hardcode the incarnation to zero, even though this function might be called with higher incarnations?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 6]
```diff
@@ -259,21 +264,41 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
...
+                    data,
+                    maybe_layout,
+                )));
+            },
+            Occupied(o) => {
+                assert!(
+                    if let EntryCell::Write(i, v, layout) = &o.get().cell {
+                        // base value may have already been provided due to a concurrency race.
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/versioned_data.rs:280]: What is concurrency race?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 7]
```diff
@@ -259,21 +264,41 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
...
+                        // If maybe_layout is None, they are required to be identical.
+                        // If maybe_layout is Some, there might have been an exchange
+                        // Assert the length of bytes for efficiency (instead of full equality)
+                        layout.is_some() == maybe_layout.is_some()
+                            && *i == 0
+                            && (layout.is_some() || v.bytes_len() == bytes_len)
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/versioned_data.rs:286]: I didn't get it. We check that the length of the data matches only when layout is None. Why is that?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 8]
```diff
@@ -259,21 +264,41 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
         let mut v = self.values.entry(key).or_default();
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/versioned_data.rs:264]: I like the name `store_base_value` or `record_based_value` better than `provide_base_value`.

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 9]
```diff
@@ -77,6 +77,30 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> Default
...
+        incarnation: Incarnation,
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:83]: The incarnation is given as input to `set_base_values` function here. But the incarnation is not given as input to `provide_base_value` function in versioned_data.rs. Any reason for that?

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 10]
```diff
@@ -77,6 +77,30 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> Default
...
+        // TODO add layout to values
+        values: impl IntoIterator<Item = (T, V)>,
+    ) {
+        match self.idx_to_update.get(&shifted_idx) {
+            Some(previous) => {
+                // base value may have already been provided due to a concurrency race.
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:89]: Please also explain here what concurrency race means.

**Review by vusirikala — APPROVED** [review]

File: aptos-move/mvhashmap/src/versioned_data.rs (hunk 5 above)
Comment (igor-aptos) [reply aptos-move/mvhashmap/src/versioned_data.rs:272]: "value" of the base value is always from version 0, and we will not re-fetch it, even if incarnation is newer, so fixed incarnation is correct here

File: aptos-move/mvhashmap/src/versioned_data.rs (hunk 7 above)
Comment (igor-aptos) [reply aptos-move/mvhashmap/src/versioned_data.rs:286]: added more comments, and explanation here - https://aptos-org.slack.com/archives/C057VPCD7QX/p1697520278313169

**Pull request merged**